src/domain/contracts/decoders/abi-decoder.helper.ts

```diff
@@ -1,78 +1,165 @@
 import {
   Abi,
+  AbiFunction,
+  AbiStateMutability,
+  ContractEventArgsFromTopics,
   ContractEventName,
+  ContractFunctionArgs,
   ContractFunctionName,
-  DecodeEventLogParameters,
-  DecodeFunctionDataParameters,
   Hex,
-  decodeEventLog as _decodeEventLog,
-  decodeFunctionData as _decodeFunctionData,
+  decodeEventLog,
+  decodeFunctionData,
   toFunctionSelector,
 } from 'viem';
 
-type Helper<TAbi extends Abi> = `is${Capitalize<ContractFunctionName<TAbi>>}`;
+export abstract class AbiDecoder<TAbi extends Abi> {
+  readonly helpers: Helpers<TAbi>;
+  readonly decodeEventLog: EventLogDecoders<TAbi>;
+  readonly decodeFunctionData: FunctionDataDecoders<TAbi>;
 
-type Helpers<TAbi extends Abi> = {
-  [key in Helper<TAbi>]: (data: Hex) => boolean;
-};
+  protected constructor(readonly abi: TAbi) {
+    this.helpers = _generateHelpers(abi);
+    this.decodeEventLog = _generateEventLogDecoders(abi);
+    this.decodeFunctionData = _generateFunctionDataDecoders(abi);
+  }
+}
 
 function capitalize<T extends string>(s: T): Capitalize<T> {
   return (s.charAt(0).toUpperCase() + s.slice(1)) as Capitalize<T>;
 }
 
+// Helpers
+
+type Helper<TAbi extends Abi> = `is${Capitalize<ContractFunctionName<TAbi>>}`;
+
+type Helpers<TAbi extends Abi> = {
+  [key in Helper<TAbi>]: (data: Hex) => boolean;
+};
+
 export function _generateHelpers<TAbi extends Abi>(
   abi: Readonly<TAbi>,
 ): Helpers<TAbi> {
+  const createHelper = (name: AbiFunction) => {
+    return (data: Hex): boolean => {
+      const functionSelector = toFunctionSelector(name);
+      return data.startsWith(functionSelector);
+    };
+  };
+
   const helpers = {} as Helpers<TAbi>;
 
   for (const item of abi) {
-    if (item.type !== 'function') {
-      continue;
+    if (item.type === 'function') {
+      const name = item.name as FunctionDataDecoder<TAbi>;
+      helpers[`is${capitalize(name)}`] = createHelper(item);
     }
+  }
+
+  return helpers;
+}
 
-    const helperName = `is${capitalize(item.name)}` as Helper<TAbi>;
-    const functionSelector = toFunctionSelector(item);
+// Function data decoders
 
-    helpers[helperName] = (data: Hex): boolean => {
-      return data.startsWith(functionSelector);
+type FunctionDataDecoder<TAbi extends Abi> = ContractFunctionName<TAbi>;
+
+type FunctionDataDecoders<TAbi extends Abi> = {
+  [key in FunctionDataDecoder<TAbi>]: (
+    data: Hex,
+  ) => ContractFunctionArgs<TAbi, AbiStateMutability, key> | null;
+};
+
+export function _generateFunctionDataDecoders<TAbi extends Abi>(
+  abi: TAbi,
+): FunctionDataDecoders<TAbi> {
+  const createFunctionDataDecoder = (name: FunctionDataDecoder<TAbi>) => {
```

Comment:
Nit: as on `_generateEventLogDecoders`, what do you think about making `createFunctionDataDecoder` an independent private function within the class?

Reply:
I adjusted it to use private functions instead in 36b45c17e2e254e6365f838b1b14e566de930810.